x/ecocredit/server/core/msg_server.go

```diff
@@ -0,0 +1,1200 @@
+		}
+		decs, err := getNonNegativeFixedDecs(precision, credit.Amount, batchSupply.TradableAmount, userBalance.Tradable, batchSupply.CancelledAmount)
+		if err != nil {
+			return nil, err
+		}
+		amtToCancelDec, supplyTradable, userBalTradable, cancelledDec := decs[0], decs[1], decs[2], decs[3]
+		userBalTradable, err = math.SafeSubBalance(userBalTradable, amtToCancelDec)
+		if err != nil {
+			return nil, err
+		}
+		supplyTradable, err = math.SafeSubBalance(supplyTradable, amtToCancelDec)
+		if err != nil {
+			return nil, err
+		}
+		cancelledDec, err = cancelledDec.Add(amtToCancelDec)
+		if err != nil {
+			return nil, err
+		}
+		if err = s.stateStore.BatchBalanceStore().Update(ctx, &ecocreditv1beta1.BatchBalance{
+			Address:  holder,
+			BatchId:  batch.Id,
+			Tradable: userBalTradable.String(),
+			Retired:  userBalance.Retired,
+		}); err != nil {
+			return nil, err
+		}
+		if err = s.stateStore.BatchSupplyStore().Update(ctx, &ecocreditv1beta1.BatchSupply{
+			BatchId:         batch.Id,
+			TradableAmount:  supplyTradable.String(),
+			RetiredAmount:   batchSupply.RetiredAmount,
+			CancelledAmount: cancelledDec.String(),
+		}); err != nil {
+			return nil, err
+		}
+		if err = sdkCtx.EventManager().EmitTypedEvent(&v1beta1.EventCancel{
+			Canceller:  holder.String(),
+			BatchDenom: credit.BatchDenom,
+			Amount:     credit.Amount,
+		}); err != nil {
+			return nil, err
+		}
+		sdkCtx.GasMeter().ConsumeGas(gasCostPerIteration, "cancel ecocredits")
+	}
+	return &v1beta1.MsgCancelResponse{}, nil
+}
+
+func (s serverImpl) UpdateClassAdmin(ctx context.Context, req *v1beta1.MsgUpdateClassAdmin) (*v1beta1.MsgUpdateClassAdminResponse, error) {
+	classInfo, err := s.stateStore.ClassInfoStore().GetByName(ctx, req.ClassId)
+	if err != nil {
+		return nil, err
+	}
+	reqAddr, _ := sdk.AccAddressFromBech32(req.Admin)
+	classAdmin := sdk.AccAddress(classInfo.Admin)
+	if !classAdmin.Equals(reqAddr) {
+		return nil, sdkerrors.ErrUnauthorized.Wrapf("expected admin %s, got %s", classInfo.Admin, req.Admin)
+	}
+	classInfo.Admin = reqAddr
+	if err = s.stateStore.ClassInfoStore().Update(ctx, classInfo); err != nil {
+		return nil, err
+	}
+	return &v1beta1.MsgUpdateClassAdminResponse{}, err
+}
+
+func (s serverImpl) UpdateClassIssuers(ctx context.Context, req *v1beta1.MsgUpdateClassIssuers) (*v1beta1.MsgUpdateClassIssuersResponse, error) {
+	class, err := s.stateStore.ClassInfoStore().GetByName(ctx, req.ClassId)
+	if err != nil {
+		return nil, err
+	}
+	reqAddr, _ := sdk.AccAddressFromBech32(req.Admin)
+	admin := sdk.AccAddress(class.Admin)
+	if !reqAddr.Equals(admin) {
+		return nil, sdkerrors.ErrUnauthorized.Wrapf("expected admin %s, got %s", class.Admin, req.Admin)
+	}
+
+	// delete the old issuers
+	if err = s.stateStore.ClassIssuerStore().DeleteBy(ctx, ecocreditv1beta1.ClassIssuerClassIdIssuerIndexKey{}.WithClassId(class.Name)); err != nil {
+		return nil, err
+	}
+
+	// add the new issuers
+	for _, issuer := range req.Issuers {
```

Comment:
it seems like `MsgUpdateClassIssuers` should have `add` and `remove` fields rather than just replacing all the old issuers

Reply:
added